File: ruby/leap/leap.rb
# Option 1: straight math
class Year
  def self.leap?(year)
    year % 4 == 0 && (year % 100 != 0 || year % 400 == 0)
  end
end

# Option 2: concept abstraction
class Year
  def self.leap?(year)
    self.div_by_4(year) && ( self.not_div_by_100(year) || self.div_by_400(year) )
  end

  def self.div_by_4(year)
    year % 4 == 0
  end

  def self.not_div_by_100(year)
    year % 100 != 0
  end

  def self.div_by_400(year)
    year % 400 == 0
  end

end


module BookKeeping
  VERSION = 3
end

File: javascript/leap/leap.js
class Year {
  constructor(input){
    this.year = input;
  }

  isLeap(){
    if ( this.divBy4(this.year) && (this.notDivBy100(this.year) || this.divBy400(this.year)) ) {
      return true;
    } else {
      return false;
    }
  }

  divBy4(input){
    return input % 4 === 0;
  }

  notDivBy100(input){
    return input % 100 != 0;
  }

  divBy400(input){
    return input % 400 === 0;
  }
}

module.exports = Year;
